### PR TITLE
Dedicated path for AWS upgrade 

### PR DESCRIPTION
**Pull request by bvtejaswi**
### :nut_and_bolt: Description: What code changed, and why?
while upgrading AWS deployment with a command, it should pull the config from FE and BE which prevents Patching the stale template config.

### :chains: Related Resources
https://chefio.atlassian.net/browse/MADROX-392

### :+1: Definition of Done
**Tested with scenarios**

- Upgraded with Aws Deployment when /hab/a2_deploy_workspace/config.toml exists , with below command

> chef-automate upgrade run --airgap-bundle latest.aib

- Upgraded with Aws Managed Services Deployment when /hab/a2_deploy_workspace/config.toml exists, with below command

> chef-automate upgrade run --airgap-bundle latest.aib --upgrade-frontends

- Upgraded with Aws Deployment when /hab/a2_deploy_workspace/config.toml  doesnot exists , with below command

> chef-automate upgrade run --airgap-bundle latest.aib
- Upgraded with Aws Managed Services Deployment when /hab/a2_deploy_workspace/config.toml doesnot exists with below command

> chef-automate upgrade run --airgap-bundle latest.aib --upgrade-frontends

### :athletic_shoe: How to Build and Test the Change
Use the following Cli  to create chef-automate utility

> bvtejaswi/automate-cli/0.1.0/20221214112309

Run Upgrade command from Bastion according to your setup
https://docs.chef.io/automate/ha_upgrade_introduction/

### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [x] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go [hunk 1]
```diff
@@ -417,6 +615,127 @@ func getHAConfigFromTFVars(tfvarConfig *HATfvars) (*ExistingInfraConfigToml, err
...
+	sharedConfigToml.Aws.Config.PostgresqlEbsVolumeIops = strings.TrimSpace(awsAutoTfvarConfig.PostgresqlEbsVolumeIops)
+	sharedConfigToml.Aws.Config.PostgresqlEbsVolumeSize = strings.TrimSpace(awsAutoTfvarConfig.PostgresqlEbsVolumeSize)
+	sharedConfigToml.Aws.Config.PostgresqlEbsVolumeType = strings.TrimSpace(awsAutoTfvarConfig.PostgresqlEbsVolumeType)
+	sharedConfigToml.Aws.Config.SetupManagedServices = awsAutoTfvarConfig.AwsSetupManagedServices
+	sharedConfigToml.Aws.Config.ChefEbsVolumeIops, _ = getTheValueFromA2HARB("chef_ebs_volume_iops")
+	return sharedConfigToml, nil
+}
+
+func getTheValueFromA2HARB(key string) (string, error) {
+	wordCountCmd := `hab pkg exec core/grep grep %s /hab/a2_deploy_workspace/a2ha.rb | wc -l`
```
Comment (punitmundra) [inline components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go:710]: hab pkg exec core/grep grep %s /hab/a2_deploy_workspace/a2ha.rb | wc -l   --> 
 HAB_LICENSE=accept-no-persist hab pkg exec core/grep grep %s /hab/a2_deploy_workspace/a2ha.rb | wc -l

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go [hunk 2]
```diff
@@ -11,6 +11,7 @@ import (
 
 	"github.com/chef/automate/components/automate-cli/pkg/status"
 	"github.com/chef/automate/components/automate-deployment/pkg/toml"
+	mtoml "github.com/chef/automate/components/automate-deployment/pkg/toml"
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:14]: is this not same as above import ?
Please use simply `toml.Marshal(sharedConfigToml)` instead of `mtoml.Marshal(sharedConfigToml)`

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go [hunk 3]
```diff
@@ -30,7 +31,30 @@ func newAwsDeployemnt(configPath string) *awsDeployment {
...
 		err := executeDeployment(args)
-		return err
+		if err != nil {
+			return err
+		}
+		sharedConfigToml, err := getAwsHAConfig()
+		if err != nil {
+			return status.Wrap(err, status.ConfigError, "unable to fetch HA config")
+		}
+		archBytes, err := ioutil.ReadFile("/hab/a2_deploy_workspace/terraform/.tf_arch") // nosemgrep
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:41]: if you want to hardcode file locations these should be put in a constant

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go [hunk 4]
```diff
@@ -30,7 +31,30 @@ func newAwsDeployemnt(configPath string) *awsDeployment {
...
+			return  err
+		}
+		var arch = strings.Trim(string(archBytes), "\n")
+		sharedConfigToml.Architecture.ConfigInitials.Architecture = arch
+		writer.Println("Reference architecture type : " + arch)
+		shardConfig, err := mtoml.Marshal(sharedConfigToml)
+		if err != nil {
+			return  status.Wrap(err, status.ConfigError, "unable to marshal config to file")
+		}
+		err = ioutil.WriteFile(filepath.Join(initConfigHabA2HAPathFlag.a2haDirPath, "config.toml"), shardConfig, 0644) // nosemgrep
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:53]: same here, please try using const for filenames if these don't need to change.

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go [hunk 5]
```diff
@@ -460,6 +815,51 @@ func getA2ORCSRootCA(config map[string]*dc.AutomateConfig) string {
...
+func getOsRoleArn(config map[string]*dc.AutomateConfig) string {
+	for _, ele := range config {
+		if ele.Global.V1.External.Opensearch != nil && ele.Global.V1.External.Opensearch.Backup != nil && ele.Global.V1.External.Opensearch.Backup.S3 != nil && ele.Global.V1.External.Opensearch.Backup.S3.Settings != nil && len(strings.TrimSpace(ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value)) > 0 {
+			return ele.Global.V1.External.Opensearch.Backup.S3.Settings.RoleArn.Value
+		}
+	}
+	return ""
+}
+
+func getOsAccessKey(config map[string]*dc.AutomateConfig) string {
```
Comment (vivek-yadav) [inline components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go:845]: Please improve unit test coverage of these functions. Have small unit tests for individual functions, so you can test all the edge cases for different types of input values.

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go [hunk 6]
```diff
@@ -321,7 +424,80 @@ func (p *PullConfigsImpl) generateConfig() (*ExistingInfraConfigToml, error) {
 	return sharedConfigToml, nil
 }
 
-func getHAConfig() (*ExistingInfraConfigToml, error) {
+func (p *PullConfigsImpl) generateAwsConfig() (*AwsConfigToml, error) {
+	sharedConfigToml, err := getAwsHAConfig()
+	if err != nil {
+		return nil, status.Wrap(err, status.ConfigError, "unable to fetch HA config")
+	}
+	archBytes, err := ioutil.ReadFile("/hab/a2_deploy_workspace/terraform/.tf_arch") // nosemgrep
```
Comment (punitmundra) [inline components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go:432]: should we remove ?

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go (hunk 3 above)
Comment (bvtejaswi) [reply components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go]: Done

File: components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go (hunk 4 above)
Comment (bvtejaswi) [reply components/automate-cli/cmd/chef-automate/automateHAAwsDeployment.go:52]: Done

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go [hunk 7]
```diff
@@ -417,6 +615,126 @@ func getHAConfigFromTFVars(tfvarConfig *HATfvars) (*ExistingInfraConfigToml, err
...
+	if err != nil {
+		return "", err
+	}
+	value := string(output)
+	count, _ := strconv.Atoi(value)
+	if count > 1 {
+		writer.Println("The key has more than one value it cannot be executed")
+		return "", nil
+	} else {
+		GrepCmd := `HAB_LICENSE=accept-no-persist hab pkg exec core/grep grep %s /hab/a2_deploy_workspace/a2ha.rb | hab pkg exec core/gawk gawk '{print $2}'`
```
[the remark this answers is not in the document]
Comment (bvtejaswi) [reply components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go]: Done

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go (hunk 5 above)
Comment (bvtejaswi) [reply components/automate-cli/cmd/chef-automate/pullAndGenerateConfig.go:845]: Done

File: components/automate-cli/pkg/testfiles/aws.auto.tfvars [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+os_snapshot_user_access_key_secret = " "
+managed_rds_instance_url = " "
+managed_rds_superuser_username = " "
+managed_rds_superuser_password = " "
+managed_rds_dbuser_username = " "
+managed_rds_dbuser_password = " "
+managed_rds_certificate = " "
+aws_ami_id = "ami-061534684251"
+lb_access_logs = "false"
+automate_lb_certificate_arn = "arn:aws:acm:us-west-1:11275563:certificate/1a0995"
```
Comment (ArvinthC3000) [inline components/automate-cli/pkg/testfiles/aws.auto.tfvars:41]: confirm if we can allow adding cert-arn in the public repo

Comment (bvtejaswi) [reply components/automate-cli/pkg/testfiles/aws.auto.tfvars:41]: Its not a valid arn

**Review by punitmundra — APPROVED** [review]

File: components/automate-cli/cmd/chef-automate/pullAndGenerateConfig_test.go [hunk 9]
```diff
@@ -10,27 +10,42 @@ import (
...
 `
+var haAwsAutoTfvarsJsonString = `
+{"backup_config_efs":"true","automate_config_file":"/hab/a2_deploy_workspace/configs/automate.toml","automate_fqdn":"A2-hello-automate-lbs-test.ap-region-1.elb.amazonaws.com","automate_instance_count":1,"chef_server_instance_count":1,"opensearch_instance_count":3,"nfs_mount_path":"/mnt/automate_backups","postgresql_instance_count":3,"postgresql_archive_disk_fs_path":"/mnt/automate_backups/postgresql","habitat_uid_gid":"","ssh_user":"test-user","ssh_port":"22","ssh_key_file":"/home/test-user/keys.pem"}
```
[the remark this answers is not in the document]
Comment (bvtejaswi) [reply components/automate-cli/cmd/chef-automate/pullAndGenerateConfig_test.go:20]: The path for automate config file was this in aws.auto.tfvars, So I have given that path
